src/Controller.php

```diff
@@ -278,12 +278,12 @@ public function getDiffs(Model $m)
             }
 
             // don't log DSQL expressions because they can be recursive and we can't store them
-            if ($original instanceof Expression || $m[$key] instanceof Expression) {
+            if ($original instanceof Expression || $m->getField($key) instanceof Expression) {
```

Comment:
Is this possible?

Reply:
i just replace with what i think is correct, i look further into it to be sure is not a useless code

Reply:
I believe should be `$m->get($key) instanceof Expression` here

Reply:
it looks for the type of Field, we already have because is get few lines before using
```php
$f = $m->hasField($key);
```
so i take directly `$f`

Reply:
@georgehristov is right that original `$m[$key]` === `$m->get($key)`

`$f = $m->hasField($key);` will be deprecated no later than today :D in https://github.com/atk4/data/pull/616 (more precisely it will be still valid, but it will return strictly bool only)

Reply:
yes, but i think was wrong even before, because in `$model->data` will be always the value, not the field type.

i'm wrong?

about hasField you are right to change the return type to boolean, is logic and correct.
i need to change even that, if not it will give error no later than today ;)

Reply:
`getField()` strictly returnds instance of `Field`. Check if `Expression` is some extended class of `Field`.

Value on the other side can be anything - because of load/save typecasting.

Reply:
that was introduced to avoid Expression to be included in audit.
this is the PR : https://github.com/atk4/audit/commit/b1e1af4ecd6e2ada0c8f8627cbe250f1237263fb#diff-3840aae95575d417f513143775141e9dR133

I clear don't remember if the value data was stored differently at that time, but now i think we have to get the model field type and compare to the Field_SQL_Expression

I'm on removing even `hasField`when is used to get the field and not to check for boolean presence

Reply:
Basically what this `if` should do is it should not audit any Expression field because expressions are not raw data - they are build based on raw data and are read-only. So no point of auditing them and also in some cases they can be recursive and then ... everything goes to hell if we try to audit them :)

Reply:
> @georgehristov is right that original `$m[$key]` === `$m->get($key)`

Expr. field will usually have field of type `Field\Callback`, but simply update the code test `$m->get($key) instanceof Expre...` like before, that makes sense.

Reply:
condition removed. Expression field has '$f->read_only === true'